woocommerce/woocommerce-android
# Issue/3725 connect errors

Parent issue #3725 

Goal of this PR is to wrap up error handling for "reader connection" flow. It introduces two new view states - ScanningFailed and ConnectingFailed.

To test:
- Unfortunately, I'm not sure how to simulate failing connection or failing scan without modifying the code.
1. Replace [this line](https://github.com/woocommerce/woocommerce-android/blob/e98aa3454fc46a06006fccd10b8f5317d92e77e8/libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/connection/ConnectionManager.kt#L43) with `CardReaderDiscoveryEvents.Failed("")`
2. Build the app
3. Open app settings
4. Tap on "Manage card reader"
5. Tap on "Redirect to CardReaderDetailFragment 
6. Tap on "Connect"
7. Notice "Scanning failed" error is shown
8. Tap on "Retry" and notice the flow is restarted (it fails again)
9. Revert changes made in `1`
10. Return "false" on [this line](https://github.com/woocommerce/woocommerce-android/blob/e98aa3454fc46a06006fccd10b8f5317d92e77e8/libs/cardreader/src/debug/java/com/woocommerce/android/cardreader/internal/connection/ConnectionManager.kt#L59)
11. Repeat steps 2-6 
12. Tap on "Connect"
13. Notice "Connecting failed" state is shown (note: It's just a simulation so the connection actually succeeds -> "retry" won't work. If you want to run some additional test you'll need to restart the app to disconnect from the reader. )

Update release notes:
